presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java

```diff
@@ -0,0 +1,423 @@
+        double retainedNdv = MoreMath.min(leftNdvInRange, rightNdvInRange);
+
+        SymbolStatsEstimate newLeftStats = buildFrom(leftStats)
+                .setNullsFraction(0)
+                .setStatisticsRange(intersect)
+                .setDistinctValuesCount(retainedNdv)
+                .build();
+
+        SymbolStatsEstimate newRightStats = buildFrom(rightStats)
+                .setNullsFraction(0)
+                .setStatisticsRange(intersect)
+                .setDistinctValuesCount(retainedNdv)
+                .build();
+
+        return stats
+                .mapSymbolColumnStatistics(clause.getLeft(), oldLeftStats -> newLeftStats)
+                .mapSymbolColumnStatistics(clause.getRight(), oldRightStats -> newRightStats)
+                .mapOutputRowCount(rowCount -> rowCount * UNKNOWN_FILTER_COEFFICIENT);
+    }
+
+    private static double firstNonNaN(double... values)
+    {
+        for (double value : values) {
+            if (!isNaN(value)) {
+                return value;
+            }
+        }
+        throw new IllegalArgumentException("All values are NaN");
+    }
+
+    @VisibleForTesting
+    JoinComplementStats calculateJoinComplementStats(
+            Optional<Expression> filter,
+            List<JoinNode.EquiJoinClause> criteria,
+            PlanNodeStatsEstimate leftStats,
+            PlanNodeStatsEstimate rightStats)
+    {
+        // TODO: add support for non-equality conditions (e.g: <=, !=, >)
+        if (filter.isPresent()) {
+            // non-equi filters are not supported
+            return new JoinComplementStats(UNKNOWN_STATS, Optional.empty());
+        }
+
+        if (criteria.isEmpty()) {
+            if (rightStats.getOutputRowCount() > 0) {
+                // all left side rows are matched
+                return new JoinComplementStats(leftStats.mapOutputRowCount(rowCount -> 0.0), Optional.empty());
+            }
+            if (rightStats.getOutputRowCount() == 0) {
+                // none left side rows are matched
+                return new JoinComplementStats(leftStats, Optional.empty());
+            }
+            // right stats row count is NaN
+            return new JoinComplementStats(UNKNOWN_STATS, Optional.empty());
+        }
+
+        // heuristics: select the most selective criteria for join complement clause
+        return IntStream.range(0, criteria.size())
+                .mapToObj(drivingClauseId -> {
+                    EquiJoinClause drivingClause = criteria.get(drivingClauseId);
+                    List<EquiJoinClause> remainingClauses = copyWithout(criteria, drivingClauseId);
+                    return calculateJoinComplementStats(leftStats, rightStats, drivingClause, remainingClauses);
+                })
+                .max(comparingDouble(joinComplementStats -> joinComplementStats.getStats().getOutputRowCount()))
+                .get();
+    }
+
+    private JoinComplementStats calculateJoinComplementStats(
```

Comment:
It would be helpful to have comments explaining the algorithm here in words (either one at the top or throughout the code).  It took me a while to get my head around what was goin on.

Reply:
@sopel39 ?